jpyinterpreter/src/main/java/org/optaplanner/jpyinterpreter/PythonExceptionTable.java

```diff
@@ -0,0 +1,176 @@
+package org.optaplanner.jpyinterpreter;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.optaplanner.jpyinterpreter.util.JumpUtils;
+
+public class PythonExceptionTable {
+    private final List<ExceptionBlock> blockList;
+
+    public PythonExceptionTable() {
+        this.blockList = new ArrayList<>();
+    }
+
+    public void addEntry(PythonVersion pythonVersion, int blockStartInstructionInclusive, int blockEndInstructionInclusive,
+            int targetByteOffset, int stackDepth, boolean pushLastIndex) {
+        blockList.add(
+                new ExceptionBlock(JumpUtils.getInstructionIndexForByteOffset(blockStartInstructionInclusive, pythonVersion),
+                        JumpUtils.getInstructionIndexForByteOffset(blockEndInstructionInclusive, pythonVersion) + 1,
+                        JumpUtils.getInstructionIndexForByteOffset(targetByteOffset, pythonVersion),
+                        stackDepth, pushLastIndex));
+    }
+
+    public List<ExceptionBlock> getEntries() {
+        return blockList;
+    }
+
+    public boolean containsJumpTarget(int target) {
+        return blockList.stream().anyMatch(block -> block.targetInstruction == target);
+    }
+
+    public Set<Integer> getJumpTargetSet() {
+        return blockList.stream().map(ExceptionBlock::getTargetInstruction).collect(Collectors.toSet());
+    }
+
+    public Set<Integer> getJumpTargetForStartSet(int start) {
+        return blockList.stream()
+                .filter(exceptionBlock -> exceptionBlock.getBlockStartInstructionInclusive() == start)
+                .map(ExceptionBlock::getTargetInstruction)
+                .collect(Collectors.toSet());
+    }
+
+    public List<ExceptionBlock> getInnerExceptionBlockList(int start, Set<Integer> possibleJumpTargetSet) {
+        return blockList.stream()
+                .filter(exceptionBlock -> exceptionBlock.getBlockStartInstructionInclusive() == start ||
+                        exceptionBlock.containsAnyTargetInSet(possibleJumpTargetSet))
+                .collect(Collectors.toList());
+    }
+
+    public Set<Integer> getStartPositionSet() {
+        return blockList.stream().map(ExceptionBlock::getBlockStartInstructionInclusive).collect(Collectors.toSet());
+    }
```

Comment:
Assumption: the interpreter is performance sensitive, and therefore we should treat its code as such. (If this assumption is incorrect because the interpreter impact is negligible, this entire comment can be disregarded.)

Thinking out loud: since we only care about `int`, all of this boxing and unboxing into `Integer` will come at a considerable cost and there are ways of avoiding it. 

If lookups aren't a big deal (perhaps because the set will be small), we can use `IntStream` and store `int[]` instead of `Set<Integer>`. If lookups are a big deal and the set is unlikely to be sparse, `BitSet` could be the way to go. Otherwise `Set<Integer>` is the only option.

Reply:
The Java bytecode generated by the interpreter is performance sensitive, since it is called many, MANY times. The code of the translator itself is not performance sensitive, it is only called once, when compiling the Python function.

There are cases where we pay a higher upfront cost in exchange for faster generated bytecode. For instance, we do typeflow and value source analysis to generate FunctionMetadata and StackMetadata. This is an expensive upfront cost, but allows us to specialize GET_ATTR from a (complicated) method call to a single GETFIELD instruction.

For the interpreter internals, clarity is far more important then performance (and this code is an interpreter internal). I think using `IntStream` or `BitSet` will add unnecessary complexity over here.